src/stream/rtsp_server.rs

```diff
@@ -0,0 +1,169 @@
+use std::sync::{Arc, Mutex};
+use std::thread;
+
+use std::collections::HashMap;
+
+use glib;
+use gstreamer_rtsp_server;
+use gstreamer_rtsp_server::prelude::{
+    RTSPMediaFactoryExt, RTSPMountPointsExt, RTSPServerExt, RTSPServerExtManual,
+};
+
+#[allow(dead_code)]
+pub struct RTSPServer {
+    pub server: gstreamer_rtsp_server::RTSPServer,
+    host: String,
+    port: u16,
+    is_running: bool,
+    pub path_to_factory: HashMap<String, gstreamer_rtsp_server::RTSPMediaFactory>,
+    main_loop_thread: Option<std::thread::JoinHandle<()>>,
+    main_loop_thread_rx_channel: std::sync::mpsc::Receiver<String>,
+}
+
+lazy_static! {
+    pub static ref RTSP_SERVER: Arc<Mutex<RTSPServer>> =
+        Arc::new(Mutex::new(RTSPServer::default()));
+}
+
+impl RTSPServer {
+    fn default() -> Self {
+        match gstreamer::init() {
+            Ok(_) => {}
+            Err(error) => println!("Error! {}", error),
+        }
+
+        let is_running = false;
+        let (sender, receiver) = std::sync::mpsc::channel::<String>();
+
+        RTSPServer {
+            server: gstreamer_rtsp_server::RTSPServer::new(),
+            host: "0.0.0.0".into(),
+            port: 8554,
+            is_running: is_running,
+            path_to_factory: HashMap::new(),
+            main_loop_thread: Some(thread::spawn(move || RTSPServer::run_main_loop(sender))),
+            main_loop_thread_rx_channel: receiver,
+        }
+    }
+
+    pub fn is_running() -> bool {
+        RTSP_SERVER.as_ref().lock().unwrap().is_running
+    }
+
+    fn run_main_loop(channel: std::sync::mpsc::Sender<String>) {
+        if let Err(error) = gstreamer::init() {
+            let _ = channel.send(format!("Failed to init GStreamer: {}", error));
+            return;
+        }
+
+        loop {
+            std::thread::sleep(std::time::Duration::from_secs(1));
+            if !RTSP_SERVER.as_ref().lock().unwrap().is_running {
+                continue;
+            }
+
+            let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();
```

Comment:
You can do the lock before the if check, using the `rtsp_server` on the if condition.

Reply:
I'm not sure if your comment is about this one  or my previous one: https://github.com/bluerobotics/mavlink-camera-manager/pull/24/files#r791807083

Reply:
Sorry, I was trying to say the following two alternatives:
```rust
        loop {
            std::thread::sleep(std::time::Duration::from_secs(1));

            let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();
            if !rtsp_server.run {
                continue;
            }
```
or
```rust
        loop {
            std::thread::sleep(std::time::Duration::from_secs(1));

            if !RTSPServer::is_running() {
                continue;
            }

            let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();

```

Reply:
The second one is being used in start()